implementations/rust/ockam/ockam_vault/src/secret_impl.rs

```diff
@@ -155,4 +188,81 @@ mod tests {
 
     #[vault_test]
     fn secret_attributes_get() {}
+
+    fn new_curve255519_attrs() -> SecretAttributes {
+        SecretAttributes::new(
+            SecretType::Curve25519,
+            SecretPersistence::Ephemeral,
+            CURVE25519_SECRET_LENGTH,
+        )
+    }
+
+    fn new_bls_attrs() -> SecretAttributes {
+        SecretAttributes::new(
+            SecretType::Bls,
+            SecretPersistence::Ephemeral,
+            BlsSecretKey::BYTES,
+        )
+    }
+
+    fn check_key_id_computation(mut vault: SoftwareVault, sec_idx: Secret) {
+        let public_key = vault.secret_public_key_get(&sec_idx).unwrap();
+        let key_id = vault.compute_key_id_for_public_key(&public_key).unwrap();
+        let sec_idx_2 = vault.get_secret_by_key_id(&key_id).unwrap();
+        assert_eq!(sec_idx, sec_idx_2)
+    }
+
+    #[test]
+    fn secret_generate_compute_key_id() {
+        for attrs in [new_curve255519_attrs(), new_bls_attrs()] {
+            let mut vault = new_vault();
+            let sec_idx = vault.secret_generate(attrs).unwrap();
+            check_key_id_computation(vault, sec_idx);
+        }
+    }
+
+    #[test]
+    fn secret_import_compute_key_id() {
+        for attrs in [new_curve255519_attrs(), new_bls_attrs()] {
+            let mut vault = new_vault();
+            let sec_idx = vault.secret_generate(attrs.clone()).unwrap();
+            let secret = vault.secret_export(&sec_idx).unwrap();
+            drop(vault); // The first vault was only used to generate random keys
+
+            let mut vault = new_vault();
+            let sec_idx = vault.secret_import(secret.as_ref(), attrs).unwrap();
+
+            check_key_id_computation(vault, sec_idx);
+        }
+    }
+
+    fn import_key(vault: &mut SoftwareVault, bytes: &[u8], attrs: SecretAttributes) -> KeyId {
+        let sec_idx = vault.secret_import(bytes, attrs).unwrap();
+        let public_key = vault.secret_public_key_get(&sec_idx).unwrap();
+        vault.compute_key_id_for_public_key(&public_key).unwrap()
+    }
+
+    #[test]
+    fn secret_import_compute_key_id_predefined() {
+        let tmp = "∇∙E=0,∇∙B=0,∇⨯E=-∂B/∂t,∇⨯B=(1/c²) ∂E/∂t".as_bytes();
+        // This works for BLS only by chance ! Use `signature_bbs_plus::SecretKey::hash` to
+        // generate a secret key for an arbitrary string.
+        let (bytes_c25519, bytes_bls): (&[u8], &[u8]) = (&tmp[0..32], &tmp[32..]);
```

Comment:
I don't think these are valid keys, so as we add key validation during import this test will fail. Let's generate some random keys for this test.
P.S. I hope Maxwell pardons me

Reply:
> Ok, I put some random (and boring ;)) keys.
> 
> Joke aside, how do I validate secret keys in ockam ?

I was trying to say that not every 32 byte binary is a valid curve25519 secret key, and I think this is true for any asymmetric cryptography (not 100% sure). Depending on curve there are some arithmetic checks needed to be done to ensure key is valid. We currently don't have them.

Reply:
> We currently don't have them.

Wouldn't that be an interesting feature to add to ockam ? I'm not sure in which crate exactly.

Reply:
@devalain yes that kind of validation would be a great addition.

Reply:
> > We currently don't have them.
> 
> Wouldn't that be an interesting feature to add to ockam ? I'm not sure in which crate exactly.

I think it goes straight to to secret_import in ockam_vault